repl: improve repl preview

BridgeAR opened the pull request:
Please check the commit messages for details.

Fixes: #33238

@nodejs/repl 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

BridgeAR commented:
@nodejs/repl PTAL. This could use some reviews.

targos commented:
I really like the behavior with the first commit!

I do not understand how the second one works, though (repl: show reference errors during preview). I tried to start node with `--use-strict`, which throws ReferenceError for bad assignments but the behavior of the REPL doesn't change.

BridgeAR commented:
@targos thanks, I did not include a check to detect `--use-strict`. If I am not mistaken, that is actually a V8 feature and not from Node.js. As such we lack any documentation of it and we implemented a own way to use the strict mode with the REPL: `NODE_REPL_MODE = 'strict'`. I am going to check if there any differences (overall they should behave identically but there might be edge cases) and think we should just listen to both and document `--use-strict`.

I am somewhat on the edge about deprecating `NODE_REPL_MODE` and to recommend `--use-strict` instead. It would only apply to "end users", so it's easy for them to switch the way the REPL is started by passing through the other flag.

BridgeAR commented:
I added support for `--use-strict` as well, since it does seem to be useful for the REPL.

BridgeAR commented:
I missed to skip the added test in case the inspector is not available. I fixed that and added another commit that improves the autocompletion by reducing the number of system calls required.

BridgeAR commented:
Landed in 47804933012841f2dc90626bdcc161adf34569a5...50ba066

MylesBorins commented:
Is this something we want to backport to v12.x? Based on the various commits referencing this PR I'm imagining it might need to be a couple different PRs all at once.

addaleax commented:
@bridgear @jasnell @targos This uses blocking operations for autocompletion. Can we revert this, or fix that? It’s extremely annoying that the repl hangs for an indefinite amount of time when you type `require('` when your `node_modules` directory is large. It also doesn’t seem necessary, given that the autocomplete API is designed to be asynchronous.

addaleax commented:
Since there’s been no reply, I’ve opened https://github.com/nodejs/node/pull/36564 to disable this by default.
